### PR TITLE
[JENKINS-41255] Propagate SCMSource id changes when the new SCMSource is identical to the old

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-41255](https://issues.jenkins-ci.org/browse/JENKINS-41255)

@reviewbybees

### CODE REVIEW

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 1]
```diff
@@ -358,6 +362,146 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
...
+                if (!removals.contains(removal.getSource().getId())) {
+                    continue;
+                }
+                if (!equalButForId(removal.getSource(), addition.getSource())) {
+                    continue;
+                }
+                changedIds.put(removal.getSource().getId(), addition.getSource().getId());
+                // now take this one out of consideration
+                removals.remove(removal.getSource().getId());
+                break;
```
Comment (rsandell) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: I'm not sure I'm following this one, you are only interested in the first removal not matching?

Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: If I have paired an addition with a removal, we don't want to check either half of the pair again

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 2]
```diff
@@ -358,6 +362,149 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
...
+        Map<String,String> changedIds = new HashMap<>();
+        Set<String> additions = new HashSet<>(newIds);
+        additions.removeAll(oldIds);
+        Set<String> removals = new HashSet<>(oldIds);
+        removals.removeAll(newIds);
+
+        // we will be manipulating the sources, so make sure we own the backing list impl
+        sources = new ArrayList<>(sources);
+        for (ListIterator<BranchSource> i = sources.listIterator(); i.hasNext(); ) {
+            BranchSource addition = i.next();
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Why not a simple

```java
for (BranchSource addition : sources) {
```

?

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 3]
```diff
@@ -358,6 +362,149 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
...
+
+    private boolean equalButForId(SCMSource a, SCMSource b) {
+        if (!a.getClass().equals(b.getClass())) {
+            return false;
+        }
+        boolean fallBack = false;
+        Class<?> clazz = a.getClass();
+        REFLECTION:
+        while (clazz != null && SCMSource.class.isAssignableFrom(clazz) && clazz != SCMSource.class) {
+            for (Field field : clazz.getDeclaredFields()) {
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Surely easier to use the XStream form for the whole `SCMSource` and just white out the `<id>…</id>`.

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 4]
```diff
@@ -358,6 +362,149 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
...
+                int modifiers = field.getModifiers();
+                if (Modifier.isStatic(modifiers)) {
+                    continue;
+                }
+                if (Modifier.isTransient(modifiers)) {
+                    continue;
+                }
+                // TODO Will Java 9 modules lock setAccessible permissions? Likely would break XStream too!
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Nothing is going to run on Java 9, forget about it…

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 5]
```diff
@@ -358,6 +362,149 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
...
+                if (aField == null || bField == null) {
+                    // we know they are different
+                    return false;
+                }
+                if (aField.equals(bField)) {
+                    // next field
+                    continue;
+                }
+                if (aField instanceof Boolean || aField instanceof Number || aField instanceof String) {
+                    // no need to compare primatives by XML, we know they are different
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: typo

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 6]
```diff
@@ -358,6 +362,149 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
...
+
+            clazz = clazz.getSuperclass();
+        }
+        return !fallBack || SOURCE_ID_OMITTED_XSTREAM.toXML(a).equals(SOURCE_ID_OMITTED_XSTREAM.toXML(b));
+    }
+
+    private final static XStream2 SOURCE_ID_OMITTED_XSTREAM = new XStream2();
+
+    static {
+        SOURCE_ID_OMITTED_XSTREAM.omitField(SCMSource.class, "id");
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: So if you can do this anyway, why the giant pile of reflection crap?

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 7]
```diff
@@ -358,6 +362,149 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
     }
 
     /**
+     * Offers direct access to set the configurable list of branch sources <strong>while</strong> preserving
+     * branch source id associations for sources that are otherwise unmodified
+     * @param sources
+     * @throws IOException
+     */
+    public void setSourcesList(List<BranchSource> sources) throws IOException {
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: If somebody wants to build an alternative to org folders but retaining multibranch, they need this method to handle legacy navigator implementations

File: src/main/java/jenkins/branch/MultiBranchProject.java (hunk 2 above)
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: I had in my mind to replace parts of the list as I iterated... for the life of me I cannot fathom why. Will change back (and remove the ArrayList 3 lines above)

File: src/main/java/jenkins/branch/MultiBranchProject.java (hunk 3 above)
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Yep. It's premature optimisation... from before I found `omitField` and was doing crazy things to ensure a bare ID would end up escaped the same way as in the whole object... I guess I'll put the reflection code back in my rainy day library

File: src/main/java/jenkins/branch/MultiBranchProject.java (hunk 6 above)
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Reflection code was fun to write... ok no it wasn't but it was less scary that the evil reflex stuff I was doing before I found omitField

**Pull request merged**